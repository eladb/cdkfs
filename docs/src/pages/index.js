import React from 'react';
import clsx from 'clsx';
import Link from '@docusaurus/Link';
import useDocusaurusContext from '@docusaurus/useDocusaurusContext';
import Layout from '@theme/Layout';

import styles from './index.module.css';

function HomepageHeader() {
  const {siteConfig} = useDocusaurusContext();
  return (
    <header className={clsx('hero hero--primary', styles.heroBanner)}>
      <div className="container">
        <h1 className="hero__title">{siteConfig.title}</h1>
        <p className="hero__subtitle">{siteConfig.tagline}</p>
        <iframe width="560" height="315" src="https://www.youtube.com/embed/SOWMPzXtTCw" title="YouTube video player"
                frameBorder="0"
                allow="accelerometer; autoplay; clipboard-write; encrypted-media; gyroscope; picture-in-picture; web-share"
                allowFullScreen></iframe>
      </div>
    </header>
  );
}

export default function Home() {
  const {siteConfig} = useDocusaurusContext();
  return (
    <Layout
      title={`Hello from ${siteConfig.title}`}
      >
      <HomepageHeader />
      <main>
      </main>
    </Layout>
  );
}
