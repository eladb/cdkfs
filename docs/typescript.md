# TypeScript Projects

To create a new TypeScript project, use `projen new typescript`:

```shell
npx projen new typescript
```

This will synthesize a standard project directory structure with some sample
code.

```shell
├── src
│   ├── index.ts
├── test
│   ├── hello.test.ts
├── node_modules
├── package.json
└── tsconfig.json
```

The default options will set up your `tsconfig` root directory as `src` and your
out directory as `lib`. By default the `test` directory is not compiled to
JavaScript 

The `projen new` command will also generate a `.projenrc.js` file which includes
the definition of your project with any options you specified in the command
line:

```js
const { typescript } = require('projen');
  const project = new typescript.TypeScriptProject({
    defaultReleaseBranch: 'main',
    name: 'my-project',
  });
project.synth();
```

> At this point, projenrc is in JavaScript, but in the future we plan to allow
> specifying your project definitions in TypeScript.

To modify your project definitions, edit `.projenrc.js` and run `projen` again
to re-synthesize your project. The following sections describe the various
features of your TypeScript project.

## Migrating your TypeScript Project to Projen

Projen provides some opinionated defaults that may be different than the defaults
you rely on in your existing TypeScript Project. The most important is the default
project structure. Projen expects all source code to be in the `src` directory and
will write compiled `.js` and `.d.ts` files to the `lib` directory. Test files are
expected in the `test` directory and not compiled to javascript. Instead, Projen
configures `Jest` to compile `.ts` files when testing.

As this is only one of many structures for valid TypeScript projects, your migrated
project may not work out-of-the-box with Projen. We recommend that you take the
time to fully migrate to the expected project structure so that you can take full
advantage of Projen management on your repository.

However, you can also configure your `.projenrc.js` file to reflect your current
project structure. A `TypeScriptProject` exposes all options in a `tsConfig` file
so you can manually set the structure you want. Here is an example for specifying a
structure where the TypeScript compiler finds all TypeScript files recursively in your
repository:

```js
const { typescript } = require('projen');
const project = new typescript.TypeScriptProject({
  ...
  tsConfig: {
    compilerOptions: {
      rootDir: '.',
      outDir: '.',
    },
    include: ['**/*.ts'], // all typescript files recursively
  }
});
```