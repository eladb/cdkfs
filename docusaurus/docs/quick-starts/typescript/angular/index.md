---
sidebar_position: 3
---

# Create an Angular Application
