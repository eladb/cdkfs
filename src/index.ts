export * from './file';
export * from './github-workflow';
export * from './ignore-file';
export * from './jsii-project';
export * from './json';
export * from './license';
export * from './node-project';
export * from './project';
export * from './component';
export * from './semver';
export * from './version';
export * from './eslint';
export * from './jest';
export * from './typescript';
export * from './mergify';
export * from './dependabot';