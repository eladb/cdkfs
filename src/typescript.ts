import * as path from 'path';
import * as fs from 'fs-extra';
import { Component } from './component';
import { Eslint } from './eslint';
import { Jest, JestOptions } from './jest';
import { JsonFile } from './json';
import { NodeProject, NodeProjectOptions } from './node-project';
import { Semver } from './semver';
import { StartEntryCategory } from './start';
import { TypedocDocgen } from './typescript-typedoc';

export interface TypeScriptProjectOptions extends NodeProjectOptions {
  /**
   * Setup jest unit tests
   * @default true
   */
  readonly jest?: boolean;

  /**
   * Jest options
   * @default - default options
   */
  readonly jestOptions?: JestOptions;

  /**
   *
   * Setup eslint.
   * @default true
   */
  readonly eslint?: boolean;

  /**
   * TypeScript version to use.
   * @default ^3.9.5
   */
  readonly typescriptVersion?: Semver;

  /**
   * Docgen by Typedoc
   *
   * @default false
   */
  readonly docgen?: boolean;

  /**
   * Docs directory
   *
   * @default 'docs'
   */
  readonly docsDirectory?: string;

  /**
   * Custom TSConfig
   *
   */
  readonly tsconfig?: TypescriptConfigOptions;

  /**
   * Do not generate a `tsconfig.json` file (used by jsii projects since
   * tsconfig.json is generated by the jsii compiler).
   *
   * @default false
   */
  readonly disableTsconfig?: boolean;

  /**
   * Compile the code before running tests.
   *
   * @default - the default behavior is to delete the lib/ directory and run
   * jest typescript tests and only if all tests pass, run the compiler.
   */
  readonly compileBeforeTest?: boolean;

  /**
   * Generate one-time sample in `src/` and `test/` if there are no files there.
   * @default true
   */
  readonly sampleCode?: boolean;

  /**
   * The .d.ts file that includes the type declarations for this module.
   * @default - .d.ts file derived from the project's entrypoint (usually lib/index.d.ts)
   */
  readonly entrypointTypes?: string;

  /**
   * Defines a `yarn package` command that will produce a tarball and place it
   * under `dist/js`.
   *
   * @default true
   */
  readonly package?: boolean;
}

/**
 * TypeScript project
 * @pjid typescript
 */
export class TypeScriptProject extends NodeProject {
  public readonly docgen?: boolean;
  public readonly docsDirectory: string;
  public readonly eslint?: Eslint;
  public readonly jest?: Jest;
  public readonly tsconfig?: TypescriptConfig;

  /**
   * The directory in which the .ts sources reside.
   */
  public readonly srcdir: string;

  /**
   * The directory in which compiled .js files reside.
   */
  public readonly libdir: string;

  /**
   * The directory in which .ts tests reside.
   */
  public readonly testdir: string;

  constructor(options: TypeScriptProjectOptions) {
    super(options);

    this.srcdir = options.srcdir ?? 'src';
    this.libdir = options.libdir ?? 'lib';
    this.testdir = options.testdir ?? 'test';

    this.docgen = options.docgen;
    this.docsDirectory = options.docsDirectory ?? 'docs/';

    this.addCompileCommand('tsc');
    this.start?.addEntry('compile', {
      desc: 'Only compile',
      category: StartEntryCategory.BUILD,
    });

    this.addScript('watch', 'tsc -w');
    this.start?.addEntry('watch', {
      desc: 'Watch & compile in the background',
      category: StartEntryCategory.BUILD,
    });

    // by default, we first run tests (jest compiles the typescript in the background) and only then we compile.
    const compileBeforeTest = options.compileBeforeTest ?? false;

    if (compileBeforeTest) {
      this.addBuildCommand(`${this.runScriptCommand} compile`, `${this.runScriptCommand} test`);
    } else {
      this.addBuildCommand(`${this.runScriptCommand} test`, `${this.runScriptCommand} compile`);
    }
    this.start?.addEntry('build', {
      desc: 'Full release build (test+compile)',
      category: StartEntryCategory.BUILD,
    });

    if (options.package ?? true) {
      this.addScript('package',
        'rm -fr dist',
        'mkdir -p dist/js',
        `${this.packageManager} pack`,
        'mv *.tgz dist/js/',
      );

      this.addBuildCommand(`${this.runScriptCommand} package`);

      this.start?.addEntry('package', {
        desc: 'Create an npm tarball',
        category: StartEntryCategory.RELEASE,
      });
    }

    if (options.entrypointTypes || this.entrypoint !== '') {
      this.manifest.types = options.entrypointTypes ?? `${path.join(path.dirname(this.entrypoint), path.basename(this.entrypoint, '.js')).replace(/\\/g, '/')}.d.ts`;
    }

    const compilerOptions = {
      alwaysStrict: true,
      declaration: true,
      experimentalDecorators: true,
      inlineSourceMap: true,
      inlineSources: true,
      lib: ['es2018'],
      module: 'CommonJS',
      noEmitOnError: false,
      noFallthroughCasesInSwitch: true,
      noImplicitAny: true,
      noImplicitReturns: true,
      noImplicitThis: true,
      noUnusedLocals: true,
      noUnusedParameters: true,
      resolveJsonModule: true,
      strict: true,
      strictNullChecks: true,
      strictPropertyInitialization: true,
      stripInternal: true,
      target: 'ES2018',
    };

    if (!options.disableTsconfig) {
      this.tsconfig = new TypescriptConfig(this, {
        include: [`${this.srcdir}/**/*.ts`],
        exclude: [
          'node_modules',
          this.libdir,
        ],
        compilerOptions: {
          rootDir: this.srcdir,
          outDir: this.libdir,
          ...compilerOptions,
        },
        ...options.tsconfig,
      });
    }

    this.gitignore.exclude(`/${this.libdir}`);
    this.npmignore?.include(`/${this.libdir}`);

    this.gitignore.include(`/${this.srcdir}`);
    this.npmignore?.exclude(`/${this.srcdir}`);

    this.npmignore?.include(`/${this.libdir}/**/*.js`);
    this.npmignore?.include(`/${this.libdir}/**/*.d.ts`);

    this.gitignore.exclude('/dist');
    this.npmignore?.exclude('dist'); // jsii-pacmak expects this to be "dist" and not "/dist". otherwise it will tamper with it

    this.npmignore?.exclude('/tsconfig.json');
    this.npmignore?.exclude('/.github');
    this.npmignore?.exclude('/.vscode');
    this.npmignore?.exclude('/.projenrc.js');

    // the tsconfig file to use for estlint (if jest is enabled, we use the jest one, otherwise we use the normal one).
    let eslintTsConfig = 'tsconfig.json';

    if (options.jest ?? true) {
      // create a tsconfig for jest that does NOT include outDir and rootDir and
      // includes both "src" and "test" as inputs.
      const tsconfig = new TypescriptConfig(this, {
        fileName: 'tsconfig.jest.json',
        include: [
          `${this.srcdir}/**/*.ts`,
          `${this.testdir}/**/*.ts`,
        ],
        exclude: [
          'node_modules',
        ],
        compilerOptions,
      });

      eslintTsConfig = tsconfig.fileName;

      // if we test before compilation, remove the lib/ directory before running
      // tests so that we get a clean slate for testing.
      if (!compileBeforeTest) {
        // make sure to delete "lib" *before* runninng tests to ensure that
        // test code does not take a dependency on "lib" and instead on "src".
        this.addTestCommand(`rm -fr ${this.libdir}/`);
      }

      this.jest = new Jest(this, {
        typescript: tsconfig,
        ...options.jestOptions,
      });

      this.gitignore.include(`/${this.testdir}`);
      this.npmignore?.exclude(`/${this.testdir}`);
    }

    if (options.eslint ?? true) {
      this.eslint = new Eslint(this, {
        tsconfigPath: `./${eslintTsConfig}`,
        dirs: [this.srcdir, this.testdir],
      });
    }

    this.addDevDependencies({
      'typescript': options.typescriptVersion ?? Semver.caret('3.9.5'),
      '@types/node': Semver.caret(this.minNodeVersion ?? '10.17.0'), // install the minimum version to ensure compatibility
    });

    // generate sample code in `src` and `lib` if these directories are empty or non-existent.
    if (options.sampleCode ?? true) {
      new SampleCode(this);
    }

    if (this.docgen) {
      new TypedocDocgen(this);
    }
  }

  /**
   * Adds commands to run as part of `yarn build`.
   * @param commands The commands to add
   */
  public addBuildCommand(...commands: string[]) {
    this.addScriptCommand('build', ...commands);
  }
}

export interface TypescriptConfigOptions {
  /**
   * @default "tsconfig.json"
   */
  readonly fileName?: string;
  /**
   * The directory in which typescript sources reside.
   * @default - all .ts files recursively
   */
  readonly include?: string[];

  /**
   * @default - node_modules is excluded by default
   */
  readonly exclude?: string[];

  /**
   * Compiler options to use.
   *
   * @default - see above
   */
  readonly compilerOptions: TypeScriptCompilerOptions;
}

export interface TypeScriptCompilerOptions {
  /**
   * Ensures that your files are parsed in the ECMAScript strict mode, and emit “use strict”
   * for each source file.
   *
   * @default true
   */
  readonly alwaysStrict?: boolean;

  /**
   * Offers a way to configure the root directory for where declaration files are emitted.
   *
   */
  readonly declarationDir?: string;

  /**
   * To be specified along with the above
   *
   */
  readonly declaration?: boolean;

  /**
   * Enables experimental support for decorators, which is in stage 2 of the TC39 standardization process.
   *
   * @default true
   */
  readonly experimentalDecorators?: boolean;

  /**
   * When set, instead of writing out a .js.map file to provide source maps,
   * TypeScript will embed the source map content in the .js files.
   *
   * @default true
   */
  readonly inlineSourceMap?: boolean;

  /**
   * When set, TypeScript will include the original content of the .ts file as an embedded
   * string in the source map. This is often useful in the same cases as inlineSourceMap.
   *
   * @default true
   */
  readonly inlineSources?: boolean;

  /**
   * Reference for type definitions / libraries to use (eg. ES2016, ES5, ES2018).
   *
   * @default [ 'es2018' ]
   */
  readonly lib?: string[];

  /**
   * Sets the module system for the program.
   * See https://www.typescriptlang.org/docs/handbook/modules.html#ambient-modules.
   *
   * @default 'CommonJS'
   */
  readonly module?: string;

  /**
   * Do not emit compiler output files like JavaScript source code, source-maps or
   * declarations if any errors were reported.
   *
   * @default true
   */
  readonly noEmitOnError?: boolean;

  /**
   * Report errors for fallthrough cases in switch statements. Ensures that any non-empty
   * case inside a switch statement includes either break or return. This means you won’t
   * accidentally ship a case fallthrough bug.
   *
   * @default true
   */
  readonly noFallthroughCasesInSwitch?: boolean;

  /**
   * In some cases where no type annotations are present, TypeScript will fall back to a
   * type of any for a variable when it cannot infer the type.
   *
   * @default true
   */
  readonly noImplicitAny?: boolean;

  /**
   * When enabled, TypeScript will check all code paths in a function to ensure they
   * return a value.
   *
   * @default true
   */
  readonly noImplicitReturns?: boolean;
  /**
   * Raise error on ‘this’ expressions with an implied ‘any’ type.
   *
   * @default true
   */
  readonly noImplicitThis?: boolean;

  /**
   * Report errors on unused local variables.
   *
   * @default true
   */
  readonly noUnusedLocals?: boolean;

  /**
   * Report errors on unused parameters in functions.
   *
   * @default true
   */
  readonly noUnusedParameters?: boolean;

  /**
   * Allows importing modules with a ‘.json’ extension, which is a common practice
   * in node projects. This includes generating a type for the import based on the static JSON shape.
   *
   * @default true
   */
  readonly resolveJsonModule?: boolean;

  /**
   * The strict flag enables a wide range of type checking behavior that results in stronger guarantees
   * of program correctness. Turning this on is equivalent to enabling all of the strict mode family
   * options, which are outlined below. You can then turn off individual strict mode family checks as
   * needed.
   *
   * @default true
   */
  readonly strict?: boolean;

  /**
   * When strictNullChecks is false, null and undefined are effectively ignored by the language.
   * This can lead to unexpected errors at runtime.
   * When strictNullChecks is true, null and undefined have their own distinct types and you’ll
   * get a type error if you try to use them where a concrete value is expected.
   *
   * @default true
   */
  readonly strictNullChecks?: boolean;

  /**
   * When set to true, TypeScript will raise an error when a class property was declared but
   * not set in the constructor.
   *
   * @default true
   */
  readonly strictPropertyInitialization?: boolean;

  /**
   * Do not emit declarations for code that has an @internal annotation in it’s JSDoc comment.
   *
   * @default true
   */
  readonly stripInternal?: boolean;

  /**
   * Modern browsers support all ES6 features, so ES6 is a good choice. You might choose to set
   * a lower target if your code is deployed to older environments, or a higher target if your
   * code is guaranteed to run in newer environments.
   *
   * @default 'ES2018'
   */
  readonly target?: string;

  /**
   * Output directory for the compiled files.
   */
  readonly outDir?: string;

  /**
   * Specifies the root directory of input files.
   *
   * Only use to control the output directory structure with `outDir`.
   */
  readonly rootDir?: string;
}

export class TypescriptConfig {
  public readonly compilerOptions: TypeScriptCompilerOptions;
  public readonly include: string[];
  public readonly exclude: string[];
  public readonly fileName: string;

  constructor(project: NodeProject, options: TypescriptConfigOptions) {
    const fileName = options.fileName ?? 'tsconfig.json';

    this.include = options.include ?? ['**/*.ts'];
    this.exclude = options.exclude ?? ['node_modules'];
    this.fileName = fileName;

    this.compilerOptions = options.compilerOptions;

    new JsonFile(project, fileName, {
      obj: {
        compilerOptions: this.compilerOptions,
        include: this.include,
        exclude: this.exclude,
      },
    });

    project.npmignore?.exclude(`/${fileName}`);
  }
}


class SampleCode extends Component {
  private readonly nodeProject: TypeScriptProject;

  constructor(project: TypeScriptProject) {
    super(project);

    this.nodeProject = project;
  }

  public synthesize(outdir: string) {
    const srcdir = path.join(outdir, this.nodeProject.srcdir);
    if (fs.pathExistsSync(srcdir) && fs.readdirSync(srcdir).filter(x => x.endsWith('.ts'))) {
      return;
    }

    const srcCode = [
      'export class Hello {',
      '  public sayHello() {',
      '    return \'hello, world!\'',
      '  }',
      '}',
    ];

    fs.mkdirpSync(srcdir);
    fs.writeFileSync(path.join(srcdir, 'index.ts'), srcCode.join('\n'));

    const testdir = path.join(outdir, this.nodeProject.testdir);
    if (fs.pathExistsSync(testdir) && fs.readdirSync(testdir).filter(x => x.endsWith('.ts'))) {
      return;
    }

    const testCode = [
      "import { Hello } from '../src'",
      '',
      "test('hello', () => {",
      "  expect(new Hello().sayHello()).toBe('hello, world!');",
      '});',
    ];

    fs.mkdirpSync(testdir);
    fs.writeFileSync(path.join(testdir, 'hello.test.ts'), testCode.join('\n'));
  }
}

/**
 * TypeScript app.
 *
 * @pjid typescript-app
 */
export class TypeScriptAppProject extends TypeScriptProject {
  constructor(options: TypeScriptProjectOptions) {
    super({
      allowLibraryDependencies: false,
      releaseWorkflow: false,
      entrypoint: '', // "main" is not needed in typescript apps
      package: false,
      ...options,
    });
  }
}

/**
 * @deprecated use `TypeScriptProject`
 */
export class TypeScriptLibraryProject extends TypeScriptProject { };

/**
 * @deprecated use TypeScriptProjectOptions
 */
export interface TypeScriptLibraryProjectOptions extends TypeScriptProjectOptions { }
