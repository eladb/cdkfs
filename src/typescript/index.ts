export * from './projenrc';
export * from './typescript';
export * from './typescript-config';
export * from './typescript-typedoc';
