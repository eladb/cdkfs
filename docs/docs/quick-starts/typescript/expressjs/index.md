---
sidebar_position: 2
---

# Create an ExpressJS Application

This quick start will help you create an [Express.js](https://expressjs.com/) application using
projen's TypeScript project. The NodeJS project is a little feature-light, so to show off
more of projen's features, we'll use the TypeScript project instead.

To begin, make sure you have the version of NodeJS you want to use set up in your terminal.
Running `which node` in bash/zsh/other POSIX shells, or `Get-Command node` in Powershell,
should print the path to the NodeJS version you want to use. [`nvm`](https://github.com/nvm-sh/nvm)
is a popular tool for managing multiple versions of Node on a single machine.
To select a version, run `nvm use <version>`.

To create a new TypeScript project, use `npx projen new typescript`:

```shell
npx projen new typescript --name=express-project
```

projen does not officially maintain a project for Express.js, so we need to install the dependencies
and write some boilerplate code.

## Dependencies

To add libraries, open `.projenrc.js` and add them to the `deps` array. For example, to add the `express` library,
add the following line to `.projenrc.ts`:

```ts
import { typescript } from "projen";

const project = new TypeScriptProject({
  name: "express-project",
  // ...
  deps: ["express"],
  // ...
});
```

Then re-run `npx projen` to install the library.

## Code

The most basic "Hello World!" example for Express.js consists of a single file, so let's create
that file and the basic app now under `src/index.ts` (replace the contents of the file that's already there):

```ts
import express, { Request, Response } from "express";
const app = express();
const port = process.env.PORT || 3000;

app.get("/", (req: Request, res: Response) => {
  console.info(req);
  res.send("Hello World!");
});

app.listen(port, () => {
  console.log(`Example app listening on port ${port}`);
});
```

## Running the app

You have code now, but you need to run it! To do that, we'll add a `start` task to projen:

```ts
project.tasks.addTask("start", {
  exec: "ts-node src/index.ts",
  receiveArgs: true,
});
```

This will run the `ts-node src/index.ts` command when you run `npx projen start`. It also makes it
possible to run `npm start`, `pnpm start` or `yarn start`, since projen updates `package.json`.
The `receiveArgs` parameter allows you to pass additional arguments to the command.

## Testing

Testing will require a few more dependencies, so let's add those now:

```ts
project.deps.addDevDeps("supertest", "@types/supertest");
```

Jest is added by default to a projen TypeScript project, so you don't need to specify it.

Next, let's update our code to make it testable:

`src/index.ts`

```ts
import express, { Request, Response } from "express";

export const exampleApp = express(); // Export the app so we can test it
const port = process.env.PORT || 3000;

exampleApp.get("/", (req: Request, res: Response) => {
  console.info(req);
  res.send("Hello World!");
});

if (require.main === module) {
  // Check if the file is being run directly
  exampleApp.listen(port, () => {
    console.log(`Example app listening on port ${port}`);
  });
}
```

Now, let's create a test file under `tests/index.test.ts` (you can delete the default test file that's already there):

```ts
import request from "supertest";
import { exampleApp } from "../src/index";

describe("GET /", () => {
  it("responds with Hello World!", async () => {
    const response = await request(exampleApp).get("/");
    expect(response.status).toBe(200);
    expect(response.text).toBe("Hello World!");
  });
});
```

projen's TypeScript project comes with a `test` task already, so we just need to run `npx projen test`.

```shell
 PASS  test/index.test.ts
  GET /
    âœ“ responds with Hello World! (23 ms)

----------|---------|----------|---------|---------|-------------------
File      | % Stmts | % Branch | % Funcs | % Lines | Uncovered Line #s
----------|---------|----------|---------|---------|-------------------
All files |   77.77 |    66.66 |      50 |   77.77 |
 index.ts |   77.77 |    66.66 |      50 |   77.77 | 12-13
----------|---------|----------|---------|---------|-------------------
Test Suites: 1 passed, 1 total
Tests:       1 passed, 1 total
Snapshots:   0 total
Time:        1.164 s
Ran all test suites.
ðŸ‘¾ test Â» eslint | eslint --ext .ts,.tsx --fix --no-error-on-unmatched-pattern src test build-tools projenrc .projenrc.ts
```

## Extending the project

While Express.js can be run from a single file, it usually is not a best practice to do so. Instead,
consider following the structure created by the [Express.js generator](https://expressjs.com/en/starter/generator.html).

Better yet, challenge yourself to [use projen to create a project for you based on the generator](/docs/custom/)!
