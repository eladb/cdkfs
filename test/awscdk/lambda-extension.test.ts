import { DependencyType, Testing } from "../../src";
import {
  AwsCdkDeps,
  AwsCdkDepsJs,
  LambdaExtension,
  LambdaRuntime,
} from "../../src/awscdk";
import { TypeScriptProject } from "../../src/typescript";

test("simplest LambdaExtension cdk v2", () => {
  const project = new TypeScriptProject({
    name: "hello",
    defaultReleaseBranch: "main",
  });

  // WHEN
  new LambdaExtension(project, {
    cdkDeps: cdkDepsForProject(project, "2.1.0"),
    entrypoint: "src/example.lambda-extension.ts",
  });

  // THEN
  const snapshot = Testing.synth(project);

  const tasks = snapshot[".projen/tasks.json"].tasks;
  const bundleTaskExec = tasks["bundle:example.lambda-extension"].steps[0].exec;

  expect(bundleTaskExec).toContain(
    // Outputs `extensions/${name}` dir
    '--outfile="assets/example.lambda-extension/extensions/example"'
  );
  expect(bundleTaskExec).toContain(
    // aws-sdk is external
    "--external:aws-sdk"
  );
  expect(bundleTaskExec).toContain(
    // Supports node12
    '--target="node12"'
  );

  const generatedSource = snapshot["src/example-layer-version.ts"];
  expect(generatedSource).toContain(
    [
      "import * as lambda from 'aws-cdk-lib/aws-lambda';",
      "import { Construct } from 'constructs';",
    ].join("\n")
  );
  expect(generatedSource).toContain(
    "export interface ExampleLayerVersionProps"
  );
  expect(generatedSource).toContain("export class ExampleLayerVersion");
  expect(generatedSource).toContain("Runtime.NODEJS_12_X");
  expect(generatedSource).toContain("Runtime.NODEJS_14_X");
  expect(generatedSource).toMatchSnapshot();
});

test("simplest LambdaExtension cdk v1", () => {
  const project = new TypeScriptProject({
    name: "hello",
    defaultReleaseBranch: "main",
  });

  // WHEN
  new LambdaExtension(project, {
    cdkDeps: cdkDepsForProject(project),
    entrypoint: "src/example.lambda-extension.ts",
  });

  // THEN
  const snapshot = Testing.synth(project);

  const generatedSource = snapshot["src/example-layer-version.ts"];
  expect(generatedSource).toContain(
    [
      "import * as lambda from '@aws-cdk/aws-lambda';",
      "import { Construct } from '@aws-cdk/core';",
    ].join("\n")
  );
});

test("changing compatible runtimes", () => {
  const project = new TypeScriptProject({
    name: "hello",
    defaultReleaseBranch: "main",
  });

  // WHEN
  new LambdaExtension(project, {
    cdkDeps: cdkDepsForProject(project),
    entrypoint: "src/example.lambda-extension.ts",
    compatibleRuntimes: [
      LambdaRuntime.NODEJS_14_X,
      LambdaRuntime.NODEJS_12_X,
      LambdaRuntime.NODEJS_10_X,
    ],
  });

  // THEN
  const snapshot = Testing.synth(project);

  const bundleTaskExec =
    snapshot[".projen/tasks.json"].tasks["bundle:example.lambda-extension"]
      .steps[0].exec;

  expect(bundleTaskExec).toContain(
    // It picked the lowest compatible runtime
    '--target="node10"'
  );

  const generatedSource = snapshot["src/example-layer-version.ts"];
  expect(generatedSource).toContain("Runtime.NODEJS_10_X");
  expect(generatedSource).toContain("Runtime.NODEJS_12_X");
  expect(generatedSource).toContain("Runtime.NODEJS_14_X");
});

test("bundler options", () => {
  const project = new TypeScriptProject({
    name: "hello",
    defaultReleaseBranch: "main",
  });

  // WHEN
  new LambdaExtension(project, {
    cdkDeps: cdkDepsForProject(project),
    entrypoint: "src/example.lambda-extension.ts",
    bundlingOptions: {
      externals: ["foo"],
    },
  });

  // THEN
  const snapshot = Testing.synth(project);

  const bundleTaskExec =
    snapshot[".projen/tasks.json"].tasks["bundle:example.lambda-extension"]
      .steps[0].exec;

  expect(bundleTaskExec).toContain(
    // `foo` is external
    "--external:foo"
  );
});

test("changing the extension name", () => {
  const project = new TypeScriptProject({
    name: "hello",
    defaultReleaseBranch: "main",
  });

  // WHEN
  new LambdaExtension(project, {
    cdkDeps: cdkDepsForProject(project),
    entrypoint: "src/example.lambda-extension.ts",
    name: "other",
  });

  // THEN
  const snapshot = Testing.synth(project);

  const bundleTaskExec =
    snapshot[".projen/tasks.json"].tasks["bundle:example.lambda-extension"]
      .steps[0].exec;

  expect(bundleTaskExec).toContain(
    // Outputs `extensions/${name}` dir
    '--outfile="assets/example.lambda-extension/extensions/other"'
  );
});

test("changing construct name and path", () => {
  const project = new TypeScriptProject({
    name: "hello",
    defaultReleaseBranch: "main",
  });

  // WHEN
  new LambdaExtension(project, {
    cdkDeps: cdkDepsForProject(project),
    entrypoint: "src/example.lambda-extension.ts",
    constructName: "Custom",
    constructFile: "src/example-extension.ts",
  });

  // THEN
  const snapshot = Testing.synth(project);
  const generatedSource = snapshot["src/example-extension.ts"];
  expect(generatedSource).toContain("export interface CustomProps");
  expect(generatedSource).toContain("export class Custom");
});

function cdkDepsForProject(
  project: TypeScriptProject,
  cdkVersion = "1.0.0"
): AwsCdkDeps {
  return new AwsCdkDepsJs(project, {
    cdkVersion: cdkVersion,
    dependencyType: DependencyType.RUNTIME,
  });
}
