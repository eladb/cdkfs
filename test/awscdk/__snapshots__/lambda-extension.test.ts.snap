// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`simplest LambdaExtension cdk v2 1`] = `
"// ~~ Generated by projen. To modify, edit .projenrc.js and run \\"npx projen\\".
import * as path from 'path';
import * as lambda from 'aws-cdk-lib/aws-lambda';
import { Construct } from 'constructs';

/**
 * Props for ExampleLayerVersion
 */
export interface ExampleLayerVersionProps extends lambda.LayerVersionOptions {
}

/**
 * An AWS Lambda Layer
 */
export class ExampleLayerVersion extends lambda.LayerVersion {
  constructor(scope: Construct, id: string, props?: ExampleLayerVersionProps) {
    super(scope, id, {
      description: '\`example\` lambda extension from src/example.lambda-extension.ts',
      ...props,
      compatibleRuntimes: [
        lambda.Runtime.NODEJS_12_X,
        lambda.Runtime.NODEJS_14_X,
      ],
      code: lambda.Code.fromAsset(path.join(__dirname,
        '../assets/example.lambda-extension')),
    });
  }
}"
`;
